src/metabase/pulse/render/body.clj

```diff
@@ -1030,16 +1030,31 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- all-unique?
+  [funnel-rows]
+  (let [ks (into #{} (map :key funnel-rows))]
+    (= (count ks) (count funnel-rows))))
+
+(defn- build-funnel-rows
+  [funnel-rows raw-rows]
+  (let [funnel-rows (map-indexed (fn [idx row] (assoc row :idx (inc idx))) funnel-rows)
+        xf-rows     (into {} (map (fn [row] (vec (take-last 2 row))) raw-rows))]
```

Comment:
this is probably a silly question, but under what circumstances would the row have more than two elements? I kinda thought in this context a row would always be an `[x, y]` pair?

Reply:
Minor note, but it's idiomatic when using `into` and `map` to move the brackets a bit and use the "transducing". This tiny adjustment avoids constructing a lazy sequence just to immediately iterate over and discard it.

```suggestion
        xf-rows     (into {} (map (fn [row] (vec (take-last 2 row)))) raw-rows)]
```

Reply:
Following on from Tim's question, what are the semantic meaning of the key and the value in this map? I think that the name of the binding should reflect that. For example if the keys represent some kind of `foo` and the values some kind of `bar`, then I would name it `foo->bar`. 

The current name trips me up as `xf` is often short for an `xform`, a transducing function, and `rows` implies some kind of sequence rather than a mapping.